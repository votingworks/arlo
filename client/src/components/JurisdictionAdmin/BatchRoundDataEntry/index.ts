import BatchRoundDataEntry from './BatchRoundDataEntry'

export default BatchRoundDataEntry
