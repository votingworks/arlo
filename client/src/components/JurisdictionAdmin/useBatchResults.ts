import { useQuery, useMutation, useQueryClient } from 'react-query'
import { fetchApi } from '../../utils/api'

export interface IBatchResults {
  [choiceId: string]: number
}

export interface IBatchResultTallySheet {
  name: string
  results: IBatchResults
}

export interface IBatch {
  id: string
  name: string
  numBallots: number
  auditBoard: null | {
    id: string
    name: string
  }
  resultTallySheets: IBatchResultTallySheet[]
}

export interface IBatches {
  batches: IBatch[]
  resultsFinalizedAt: string | null
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useBatches = (
  electionId: string,
  jurisdictionId: string,
  roundId: string
) =>
  useQuery<IBatches>('batches', () =>
    fetchApi(
      `/api/election/${electionId}/jurisdiction/${jurisdictionId}/round/${roundId}/batches`
    )
  )

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useRecordBatchResults = (
  electionId: string,
  jurisdictionId: string,
  roundId: string
) => {
  const putBatchResults = async ({
    batchId,
    resultTallySheets,
  }: {
    batchId: string
    resultTallySheets: IBatchResultTallySheet[]
  }) =>
    fetchApi(
      `/api/election/${electionId}/jurisdiction/${jurisdictionId}/round/${roundId}/batches/${batchId}/results`,
      {
        method: 'PUT',
        body: JSON.stringify(resultTallySheets),
        headers: { 'Content-Type': 'application/json' },
      }
    )

  const queryClient = useQueryClient()

  return useMutation(putBatchResults, {
    onSuccess: () => queryClient.invalidateQueries('batches'),
  })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useFinalizeBatchResults = (
  electionId: string,
  jurisdictionId: string,
  roundId: string
) => {
  const finalizeBatchResults = async () =>
    fetchApi(
      `/api/election/${electionId}/jurisdiction/${jurisdictionId}/round/${roundId}/batches/finalize`,
      { method: 'POST' }
    )

  const queryClient = useQueryClient()

  return useMutation(finalizeBatchResults, {
    onSuccess: () => queryClient.invalidateQueries('batches'),
  })
}
