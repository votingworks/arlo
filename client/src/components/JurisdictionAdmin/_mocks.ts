import { IBatches } from './useBatchResults'
import { IRound } from '../AuditAdmin/useRoundsAuditAdmin'
import { FileProcessingStatus } from '../useCSV'
import {
  IFullHandTallyBatchResults,
  IFullHandTallyBatchResult,
} from './useFullHandTallyResults'
import { mocksOfType } from '../testUtilities'

export interface INullResultValues {
  [contestId: string]: {
    [choiceId: string]: null | string | number
  }
}

export const roundMocks = mocksOfType<IRound>()({
  incomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: false,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  complete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: '2020-09-14T17:35:19.482Z',
    needsFullHandTally: false,
    isAuditComplete: true,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  fullHandTallyIncomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: true,
    isFullHandTally: true,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
})

export const resultsMocks = mocksOfType<INullResultValues>()({
  emptyInitial: {
    'contest-id-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'contest-id-2': {
      'choice-id-3': null,
      'choice-id-4': null,
    },
  },
  complete: {
    'contest-id-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'contest-id-2': {
      'choice-id-3': 1,
      'choice-id-4': 2,
    },
  },
})

export const batchResultsMocks = mocksOfType<INullResultValues>()({
  empty: {
    'batch-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-2': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-3': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
  },
  complete: {
    'batch-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-2': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-3': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
  },
})

export const batchesMocks = mocksOfType<IBatches>()({
  emptyInitial: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
    ],
    resultsFinalizedAt: null,
  },
  complete: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 1,
              'choice-id-2': 2,
            },
          },
        ],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 0,
              'choice-id-2': 10,
            },
          },
        ],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 2000,
              'choice-id-2': 20,
            },
          },
        ],
      },
    ],
    resultsFinalizedAt: '2020-09-14T17:35:19.482Z',
  },
})

export const fullHandTallyBatchResultMock = mocksOfType<
  IFullHandTallyBatchResults
>()({
  empty: {
    finalizedAt: '',
    results: [],
  },
  complete: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  completeWithMultipleBatch: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 15,
        },
      },
      {
        batchName: 'Batch2',
        batchType: 'Provisional',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 5,
        },
      },
    ],
  },
  updated: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  finalized: {
    finalizedAt: '2021-04-13T15:01:00.383031+00:00',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
})

export const fullHandTallyBatchResultsMock = mocksOfType<
  IFullHandTallyBatchResult
>()({
  empty: {
    batchName: '',
    batchType: '',
    choiceResults: {},
  },
  complete: {
    batchName: 'Batch1',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
  updated: {
    batchName: 'Batch12',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
})
