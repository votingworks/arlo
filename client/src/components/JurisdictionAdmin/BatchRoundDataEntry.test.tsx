import React from 'react'
import { screen, within, waitFor } from '@testing-library/react'
import { useParams } from 'react-router-dom'
import { QueryClientProvider } from 'react-query'
import userEvent from '@testing-library/user-event'
import copy from 'copy-to-clipboard'
import { ToastContainer } from 'react-toastify'
import BatchRoundDataEntry from './BatchRoundDataEntry'
import { batchesMocks } from './_mocks'
import { IBatches, IBatchResultTallySheet } from './useBatchResults'
import { IContest } from '../../types'
import {
  renderWithRouter,
  withMockFetch,
  findAndCloseToast,
  serverError,
} from '../testUtilities'
import { queryClient } from '../../App'
import {
  contestMocks,
  roundMocks,
} from '../AuditAdmin/useSetupMenuItems/_mocks'

jest.mock('copy-to-clipboard', () => jest.fn(() => true))

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useRouteMatch: jest.fn(),
  useParams: jest.fn(),
}))
const paramsMock = useParams as jest.Mock
paramsMock.mockReturnValue({
  electionId: '1',
  jurisdictionId: '1',
})

const apiCalls = {
  getJAContests: (response: { contests: IContest[] }) => ({
    url: `/api/election/1/jurisdiction/1/contest`,
    response,
  }),
  getBatches: (response: IBatches) => ({
    url: '/api/election/1/jurisdiction/1/round/round-1/batches',
    response,
  }),
  putBatchResults: (
    batchId: string,
    resultTallySheets: IBatchResultTallySheet[]
  ) => ({
    url: `/api/election/1/jurisdiction/1/round/round-1/batches/${batchId}/results`,
    options: {
      method: 'PUT',
      body: JSON.stringify(resultTallySheets),
      headers: {
        'Content-Type': 'application/json',
      },
    },
    response: { status: 'ok' },
  }),
  finalizeBatchResults: () => ({
    url: `/api/election/1/jurisdiction/1/round/round-1/batches/finalize`,
    options: { method: 'POST' },
    response: { status: 'ok' },
  }),
}

const batchesWithResults = (resultTallySheets: IBatchResultTallySheet[]) => [
  {
    ...batchesMocks.emptyInitial.batches[0],
    resultTallySheets,
  },
  ...batchesMocks.emptyInitial.batches.slice(1),
]

const render = () =>
  renderWithRouter(
    <QueryClientProvider client={queryClient}>
      <BatchRoundDataEntry round={roundMocks.singleIncomplete[0]} />
      <ToastContainer />
    </QueryClientProvider>,
    {
      route: '/election/1/jurisdiction/1',
    }
  )

describe('Batch comparison data entry', () => {
  it('shows a table of batches and a button to edit results for each batch', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches(batchesMocks.emptyInitial),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const batchTable = await screen.findByRole('table')

      const headers = within(batchTable).getAllByRole('columnheader')
      expect(headers).toHaveLength(5)
      expect(headers[0]).toHaveTextContent('Batch Name')
      expect(headers[1]).toHaveTextContent('Choice One')
      expect(headers[2]).toHaveTextContent('Choice Two')
      expect(headers[3]).toHaveTextContent('Batch Total Votes')
      expect(headers[4]).toHaveTextContent('Actions')

      const rows = within(batchTable).getAllByRole('row')
      expect(rows).toHaveLength(5)

      const row1 = within(rows[1]).getAllByRole('cell')
      expect(row1).toHaveLength(5)
      expect(row1[0]).toHaveTextContent('Batch One')
      expect(row1[1]).toBeEmpty()
      expect(row1[2]).toBeEmpty()
      expect(row1[3]).toBeEmpty()
      within(row1[4]).getByRole('button', { name: /Edit/ })

      const row2 = within(rows[2]).getAllByRole('cell')
      expect(row2).toHaveLength(5)
      expect(row2[0]).toHaveTextContent('Batch Two')
      expect(row2[1]).toBeEmpty()
      expect(row2[2]).toBeEmpty()
      expect(row2[3]).toBeEmpty()
      within(row2[4]).getByRole('button', { name: /Edit/ })

      const row3 = within(rows[3]).getAllByRole('cell')
      expect(row3).toHaveLength(5)
      expect(row3[0]).toHaveTextContent('Batch Three')
      expect(row3[1]).toBeEmpty()
      expect(row3[2]).toBeEmpty()
      expect(row3[3]).toBeEmpty()
      within(row3[4]).getByRole('button', { name: /Edit/ })

      const row4 = within(rows[4]).getAllByRole('cell')
      expect(row4).toHaveLength(5)
      expect(row4[0]).toHaveTextContent('Choice Total Votes')
      expect(row4[1]).toHaveTextContent('0')
      expect(row4[2]).toHaveTextContent('0')
      expect(row4[3]).toHaveTextContent('0')
      expect(row4[4]).toBeEmpty()
    })
  })

  it('edits the results for a batch', async () => {
    const tallySheet1 = {
      name: 'Tally Sheet #1',
      results: {
        'choice-id-1': 1,
        'choice-id-2': 2,
      },
    }
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches(batchesMocks.emptyInitial),
      apiCalls.putBatchResults('batch-1', [tallySheet1]),
      apiCalls.getBatches({
        ...batchesMocks.emptyInitial,
        batches: batchesWithResults([tallySheet1]),
      }),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const batchTable = await screen.findByRole('table')
      let rows = within(batchTable).getAllByRole('row')
      userEvent.click(within(rows[1]).getByRole('button', { name: /Edit/ }))

      const saveButton = await screen.findByRole('button', {
        name: /Save/,
      })
      rows = within(batchTable).getAllByRole('row')
      let row1 = within(rows[1]).getAllByRole('cell')

      const choice1Input = within(row1[1]).getByRole('spinbutton')
      userEvent.type(choice1Input, '1')
      await waitFor(
        () => expect(row1[3]).toHaveTextContent('1') // Batch total should update
      )

      // Try saving before filling out all choices
      userEvent.click(saveButton)

      // Should get a validation error
      const choice2Input = within(row1[2]).getByRole('spinbutton')
      await waitFor(() =>
        expect(choice2Input).toHaveStyle({ borderColor: 'red' })
      )
      expect(choice2Input).toHaveFocus()

      // Fill out the rest of the choices
      userEvent.type(choice2Input, '2')
      await waitFor(
        () => expect(row1[3]).toHaveTextContent('3') // Batch total should update
      )

      userEvent.click(saveButton)

      await waitFor(() => expect(saveButton).not.toBeInTheDocument())
      rows = within(batchTable).getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      within(row1[4]).getByRole('button', { name: /Edit/ })
      expect(row1[1]).toHaveTextContent('1')
      expect(row1[2]).toHaveTextContent('2')
      expect(row1[3]).toHaveTextContent('3')

      const row4 = within(rows[4]).getAllByRole('cell')
      expect(row4[1]).toHaveTextContent('1')
      expect(row4[2]).toHaveTextContent('2')
      expect(row4[3]).toHaveTextContent('3')

      // Test copy button
      const copyButton = screen.getByRole('button', {
        name: /Copy to clipboard/,
      })
      userEvent.click(copyButton)

      expect(copy).toHaveBeenCalled()
      expect((copy as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
    })
  })

  it('cancels editing the results for a batch', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches(batchesMocks.emptyInitial),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const batchTable = await screen.findByRole('table')
      let rows = within(batchTable).getAllByRole('row')
      userEvent.click(within(rows[1]).getByRole('button', { name: /Edit/ }))

      const cancelButton = await screen.findByRole('button', {
        name: /Cancel/,
      })
      rows = within(batchTable).getAllByRole('row')
      let row1 = within(rows[1]).getAllByRole('cell')

      const choice1Input = within(row1[1]).getByRole('spinbutton')
      userEvent.type(choice1Input, '1')
      const choice2Input = within(row1[2]).getByRole('spinbutton')
      userEvent.type(choice2Input, '2')

      userEvent.click(cancelButton)
      await waitFor(() => expect(cancelButton).not.toBeInTheDocument())

      rows = within(batchTable).getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      expect(row1[1]).toBeEmpty()
      expect(row1[2]).toBeEmpty()
      expect(row1[3]).toBeEmpty()
      within(row1[4]).getByRole('button', { name: /Edit/ })
    })
  })

  it('edits multiple tally sheets for a batch', async () => {
    const tallySheet1 = {
      name: 'Tally Sheet #1',
      results: {
        'choice-id-1': 1,
        'choice-id-2': 2,
      },
    }
    const tallySheet2 = {
      name: 'Second Tally Sheet',
      results: {
        'choice-id-1': 3,
        'choice-id-2': 4,
      },
    }
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches(batchesMocks.emptyInitial),
      apiCalls.putBatchResults('batch-1', [tallySheet1, tallySheet2]),
      apiCalls.getBatches({
        ...batchesMocks.emptyInitial,
        batches: batchesWithResults([tallySheet1, tallySheet2]),
      }),
      apiCalls.putBatchResults('batch-1', [tallySheet2]),
      apiCalls.getBatches({
        ...batchesMocks.emptyInitial,
        batches: batchesWithResults([tallySheet2]),
      }),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const batchTable = await screen.findByRole('table')
      let rows = within(batchTable).getAllByRole('row')
      userEvent.click(within(rows[1]).getByRole('button', { name: /More/ }))
      userEvent.click(screen.getByText('Use Multiple Tally Sheets'))

      let dialog = (
        await screen.findByRole('heading', {
          name: /Edit Tally Sheets for Batch: Batch One/,
        })
      ).closest('.bp3-dialog')! as HTMLElement
      let tallySheetsTable = within(dialog).getByRole('table')

      const headers = within(tallySheetsTable).getAllByRole('columnheader')
      expect(headers).toHaveLength(4)
      expect(headers[0]).toHaveTextContent('Tally Sheet Label')
      expect(headers[1]).toHaveTextContent('Choice One')
      expect(headers[2]).toHaveTextContent('Choice Two')
      expect(headers[3]).toHaveTextContent('Actions')

      rows = within(tallySheetsTable).getAllByRole('row')
      expect(rows).toHaveLength(2)

      let row1 = within(rows[1]).getAllByRole('cell')
      expect(row1).toHaveLength(4)
      expect(within(row1[0]).getByRole('textbox')).toHaveValue('Tally Sheet #1')
      const choice1Input = within(row1[1]).getByRole('spinbutton')
      expect(choice1Input).toBeEmpty()
      const choice2Input = within(row1[2]).getByRole('spinbutton')
      expect(choice2Input).toBeEmpty()
      expect(
        within(row1[3]).getByRole('button', { name: /Delete/ })
      ).toBeDisabled()

      // Fill out one choice
      userEvent.type(choice1Input, '1')

      // Try saving before filling out all choices
      userEvent.click(screen.getByRole('button', { name: /Save/ }))

      // Should get a validation error
      await waitFor(() =>
        expect(choice2Input).toHaveStyle({ borderColor: 'red' })
      )
      expect(choice2Input).toHaveFocus()

      // Fill out the other choice
      userEvent.type(choice2Input, '2')

      // Add another tally sheet
      userEvent.click(screen.getByRole('button', { name: /Add Tally Sheet/ }))
      rows = within(tallySheetsTable).getAllByRole('row')
      const row2 = within(rows[2]).getAllByRole('cell')

      // Delete the default label
      const labelInput = within(row2[0]).getByRole('textbox')
      expect(labelInput).toHaveValue('Tally Sheet #2')
      userEvent.clear(labelInput)

      // Fill out the choices
      userEvent.type(within(row2[1]).getByRole('spinbutton'), '3')
      userEvent.type(within(row2[2]).getByRole('spinbutton'), '4')

      // Try saving without tally sheet label
      userEvent.click(screen.getByRole('button', { name: /Save/ }))

      // Should get a validation error
      await waitFor(() =>
        expect(labelInput).toHaveStyle({ borderColor: 'red' })
      )
      expect(labelInput).toHaveFocus()

      // Enter a label
      userEvent.type(labelInput, 'Second Tally Sheet')

      // Now save - should show in main results table
      userEvent.click(screen.getByRole('button', { name: /Save/ }))
      await waitFor(() => expect(dialog).not.toBeInTheDocument())
      rows = within(batchTable).getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      // Choice tally sheets should be summed
      expect(row1[1]).toHaveTextContent(`${1 + 3}`)
      expect(row1[2]).toHaveTextContent(`${2 + 4}`)
      expect(row1[3]).toHaveTextContent(`${1 + 3 + 2 + 4}`)
      const editTallySheetsButton = within(row1[4]).getByRole('button', {
        name: /Edit Tally Sheets/,
      })
      expect(
        within(row1[4]).queryByRole('button', { name: /More/ })
      ).not.toBeInTheDocument()

      // Reopen the tally sheets modal
      userEvent.click(editTallySheetsButton)
      dialog = (
        await screen.findByRole('heading', {
          name: /Edit Tally Sheets for Batch: Batch One/,
        })
      ).closest('.bp3-dialog')! as HTMLElement
      tallySheetsTable = within(dialog).getByRole('table')
      rows = within(tallySheetsTable).getAllByRole('row')
      expect(rows).toHaveLength(3)

      // Delete the first tally sheet
      within(rows[1])
        .getByRole('button', { name: /Delete/ })
        .click()
      rows = within(tallySheetsTable).getAllByRole('row')
      expect(rows).toHaveLength(2)

      // Save the changes - should revert main results table to a single tally sheet
      userEvent.click(screen.getByRole('button', { name: /Save/ }))
      await waitFor(() => expect(dialog).not.toBeInTheDocument())
      rows = within(batchTable).getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      expect(row1[1]).toHaveTextContent('3')
      expect(row1[2]).toHaveTextContent('4')
      expect(row1[3]).toHaveTextContent(`${3 + 4}`)
      within(row1[4]).getByRole('button', {
        name: /Edit/,
      })
      const moreButton = within(row1[4]).getByRole('button', { name: /More/ })

      // Open again
      userEvent.click(moreButton)
      userEvent.click(screen.getByText('Use Multiple Tally Sheets'))
      dialog = (
        await screen.findByRole('heading', {
          name: /Edit Tally Sheets for Batch: Batch One/,
        })
      ).closest('.bp3-dialog')! as HTMLElement
      tallySheetsTable = within(dialog).getByRole('table')
      rows = within(tallySheetsTable).getAllByRole('row')
      expect(rows).toHaveLength(3)
      expect(within(rows[2]).getByRole('textbox')).toHaveValue('Tally Sheet #2')

      // Cancel - no changes in main table
      userEvent.click(screen.getByRole('button', { name: /Cancel/ }))
      await waitFor(() => expect(dialog).not.toBeInTheDocument())
      rows = within(batchTable).getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      expect(row1[1]).toHaveTextContent('3')
      expect(row1[2]).toHaveTextContent('4')
      expect(row1[3]).toHaveTextContent(`${3 + 4}`)
      within(row1[4]).getByRole('button', {
        name: /Edit/,
      })
      within(row1[4]).getByRole('button', { name: /More/ })
    })
  })

  it('finalizes results', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches({
        ...batchesMocks.complete,
        resultsFinalizedAt: null,
      }),
      apiCalls.finalizeBatchResults(),
      apiCalls.getBatches(batchesMocks.complete),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const finalizeButton = await screen.findByRole('button', {
        name: /Finalize Results/,
      })
      userEvent.click(finalizeButton)

      const dialog = (
        await screen.findByRole('heading', {
          name: 'Are you sure you want to finalize your results?',
        })
      ).closest('.bp3-dialog')! as HTMLElement
      userEvent.click(within(dialog).getByRole('button', { name: /Confirm/ }))

      await screen.findByText('Results finalized')
      expect(finalizeButton).toBeDisabled()
      const editButtons = screen.getAllByRole('button', { name: /Edit/ })
      editButtons.forEach(button => expect(button).toBeDisabled())
      expect(
        screen.getByRole('button', { name: /Copy to clipboard/ })
      ).toBeEnabled()
    })
  })

  it('disallows finalizing until all batches have results', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches({
        batches: [
          batchesMocks.emptyInitial.batches[0],
          ...batchesMocks.complete.batches.slice(1),
        ],
        resultsFinalizedAt: null,
      }),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const finalizeButton = await screen.findByRole('button', {
        name: /Finalize Results/,
      })
      userEvent.click(finalizeButton)

      await findAndCloseToast(
        'Please enter results for all batches before finalizing.'
      )
    })
  })

  it('handles errors on save', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches(batchesMocks.emptyInitial),
      serverError(
        'putBatchResults',
        apiCalls.putBatchResults('batch-1', [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 1,
              'choice-id-2': 2,
            },
          },
        ])
      ),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      const batchTable = await screen.findByRole('table')
      let rows = within(batchTable).getAllByRole('row')
      userEvent.click(within(rows[1]).getByRole('button', { name: /Edit/ }))

      const saveButton = await screen.findByRole('button', {
        name: /Save/,
      })
      rows = within(batchTable).getAllByRole('row')
      const row1 = within(rows[1]).getAllByRole('cell')

      const choice1Input = within(row1[1]).getByRole('spinbutton')
      userEvent.type(choice1Input, '1')
      const choice2Input = within(row1[2]).getByRole('spinbutton')
      userEvent.type(choice2Input, '2')

      userEvent.click(saveButton)

      await findAndCloseToast('something went wrong: putBatchResults')
      expect(saveButton).toBeInTheDocument()
    })
  })

  it('handles errors on finalize', async () => {
    const expectedCalls = [
      apiCalls.getJAContests({ contests: contestMocks.oneTargeted }),
      apiCalls.getBatches({
        ...batchesMocks.complete,
        resultsFinalizedAt: null,
      }),
      serverError('finalizeBatchResults', apiCalls.finalizeBatchResults()),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      userEvent.click(
        await screen.findByRole('button', { name: /Finalize Results/ })
      )
      const dialog = (
        await screen.findByRole('heading', {
          name: 'Are you sure you want to finalize your results?',
        })
      ).closest('.bp3-dialog')! as HTMLElement
      userEvent.click(within(dialog).getByRole('button', { name: /Confirm/ }))
      await findAndCloseToast('something went wrong: finalizeBatchResults')
    })
  })
})
