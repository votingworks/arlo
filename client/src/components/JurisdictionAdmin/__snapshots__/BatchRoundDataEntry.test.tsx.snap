// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batch comparison data entry edits the results for a batch 1`] = `"<table id=\\"results-table\\" class=\\"bp3-html-table bp3-html-table-bordered bp3-html-table-striped sc-bdVaJa lhFUpu\\"><thead><tr><th style=\\"width: 22%;\\">Batch Name</th><th>Choice One</th><th>Choice Two</th><th class=\\"sc-bxivhb logWDV\\">Batch Total Votes</th><th style=\\"width: 120px;\\">Actions</th></tr></thead><tbody><tr><td>Batch One</td><td class=\\"sc-bwzfXH jazXpl\\">1</td><td class=\\"sc-bwzfXH jazXpl\\">2</td><td class=\\"sc-htpNat dSZuaR\\">3</td><td><button type=\\"button\\" class=\\"bp3-button\\"><span icon=\\"edit\\" class=\\"bp3-icon bp3-icon-edit\\"><svg data-icon=\\"edit\\" width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\"><desc>edit</desc><path d=\\"M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z\\" fill-rule=\\"evenodd\\"></path></svg></span><span class=\\"bp3-button-text\\">Edit</span></button></td></tr><tr><td>Batch Two</td><td class=\\"sc-bwzfXH jazXpl\\"></td><td class=\\"sc-bwzfXH jazXpl\\"></td><td class=\\"sc-htpNat dSZuaR\\"></td><td><button type=\\"button\\" class=\\"bp3-button\\"><span icon=\\"edit\\" class=\\"bp3-icon bp3-icon-edit\\"><svg data-icon=\\"edit\\" width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\"><desc>edit</desc><path d=\\"M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z\\" fill-rule=\\"evenodd\\"></path></svg></span><span class=\\"bp3-button-text\\">Edit</span></button></td></tr><tr><td>Batch Three</td><td class=\\"sc-bwzfXH jazXpl\\"></td><td class=\\"sc-bwzfXH jazXpl\\"></td><td class=\\"sc-htpNat dSZuaR\\"></td><td><button type=\\"button\\" class=\\"bp3-button\\"><span icon=\\"edit\\" class=\\"bp3-icon bp3-icon-edit\\"><svg data-icon=\\"edit\\" width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 16 16\\"><desc>edit</desc><path d=\\"M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z\\" fill-rule=\\"evenodd\\"></path></svg></span><span class=\\"bp3-button-text\\">Edit</span></button></td></tr><tr><td class=\\"sc-htpNat dSZuaR\\">Choice Total Votes</td><td class=\\"sc-htpNat dSZuaR\\">1</td><td class=\\"sc-htpNat dSZuaR\\">2</td><td class=\\"sc-htpNat dSZuaR\\">3</td><td></td></tr></tbody></table>"`;
